modules/text-input/react/lib/Combobox.tsx

```diff
@@ -0,0 +1,265 @@
+    const lastItem = autoCompleteItemCount - 1;
+    let nextIndex = null;
+
+    switch (event.key) {
+      case 'ArrowUp':
+        const upIndex = currentIndex != null ? currentIndex - 1 : lastItem;
+        nextIndex = upIndex < 0 ? lastItem : upIndex;
+        break;
+
+      case 'ArrowDown':
+        const downIndex = currentIndex != null ? currentIndex + 1 : firstItem;
+        nextIndex = downIndex >= autoCompleteItemCount ? firstItem : downIndex;
+        break;
+
+      case 'Escape':
+        this.resetSearchInput();
+        break;
+
+      case 'Enter':
+        if (this.state.selectedAutocompleteIndex != null) {
+          const item = this.props.autocompleteItems[this.state.selectedAutocompleteIndex];
+          this.handleAutocompleteClick(event, item.props);
+          event.stopPropagation();
+          event.preventDefault();
+        }
+        break;
+
+      default:
+    }
+    this.setState({ selectedAutocompleteIndex: nextIndex });
+  };
+
+  handleSearchInputChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
+    if (this.props.onChange) {
+      this.props.onChange(event)
+    }
+    this.setState({ value: event.target.value });
+  };
+
+  render() {
+    const {
+      accessibleId,
+      autocompleteItems,
+      grow,
+      hiddenLabel,
+      onChange,
+      placeholder,
+      type,
+      ...elemProps
+    } = this.props;
+    const listBoxIdPart = `listbox`;
+    const labelIdPart = `label`;
+    const optionIdPart = `option`;
+
+    return (
+      <Container
+        grow={grow}
+        role='combobox'
+        aria-haspopup='listbox'
+        aria-owns={`${accessibleId}-${listBoxIdPart}`}
+        aria-expanded={autocompleteItems && autocompleteItems.length > 0}
+        innerRef={this.comboboxRef}
+        {...elemProps}
+      >
+        <Status role="status" aria-live="polite">
+          {autocompleteItems ? this.state.showingAutocomplete && this.buildStatusString(autocompleteItems.length) : ''}
+        </Status>
+        <FormField label="Label" inputId={`${accessibleId}-${labelIdPart}`} hiddenLabel={hiddenLabel}>
+          <TextInput
+            grow={grow}
+            value={this.state.value}
+            placeholder={placeholder}
+            type={type}
+            inputRef={this.inputRef}
+            id={`${accessibleId}-${labelIdPart}`}
+            aria-autocomplete='list'
+            aria-controls={`${accessibleId}-${listBoxIdPart}`}
+            aria-activedescendant={this.state.selectedAutocompleteIndex != null ? `${accessibleId}-${optionIdPart}-${this.state.selectedAutocompleteIndex}` : ''}
+            onChange={this.handleSearchInputChange}
+            onKeyDown={this.handleKeyboardShortcuts}
+            onFocus={this.setFocused}
+            onBlur={this.setFocused}
+          />
+        </FormField>
+        {
+          this.state.showingAutocomplete && (
+            <MenuContainer padding={spacing.zero} depth={depth[1]}>
+              <Autocomplete
```

Comment:
That's more of an overall design decision and not something addressed in this PR. It was just something I noticed that spurred the question with a specific example.

I don't think 2 lines vs 1 line change in a diff if we want to switch and element from a span to a div matters much.

Code is read more often than it is written and by people that don't share the same context as the developer who wrote it (when they wrote it). The abstraction is offering very little value and just requires cognitive overhead of another (and often many) lookups. Also if I'm debugging components, that abstraction is harmful. This is what it looks like in my React dev tools:

![Screen Shot 2019-08-29 at 4 18 00 PM](https://user-images.githubusercontent.com/338257/63980387-aa168680-ca78-11e9-835b-c8c94091b84b.png)

I now have to manually map "div `Styled`" to a randomly chosen component name.

I'm not a huge fan of the Emotion 10 API requiring a JSX pragma change to add `css=`, but it is very readable. Using the `styled` API, I have a hard time telling if `Autocomplete` is a Component or an element with styling applied. I have to jump sometimes hundreds of lines around a file to understand a render function. Doable? Sure. It just requires me to understand multiple abstraction levels simultaneously which is an indicator of a bad abstraction. Maybe I just haven't drank enough of the styled-components kool-aid yet.

I wouldn't necessarily change anything in this PR except maybe the name of the variable holding the styled element (`AutocompleteList`?). We'll bring it up with the team. I've found sometimes decisions are made arbitrarily and others with a specific set of requirements - it just takes digging.

Reply:
Totally understandable. Will change to AutocompleteList. 
Curious to hear where y'all will land on this.
I wonder if it could be something as simple as prepending styled components with the word "Styled".
e.g.
`StyledAutocompleteList`, that would communicate that it is an emotion component and wouldn't risk a mismatch like `StyledUl` ending up pointing to a `dl`